File: src/shared/telemetry/flight-path.ts
/**
 * Pure helpers to turn parsed DJI telemetry into a flight path — no DOM, no
 * map library. Extracts the GPS track from the cues the SRT parser already
 * produced, so the same data that drives the Telemetry panels draws the map.
 */

import type { Cue } from './srt-parser';

export interface TrackPoint {
  lon: number;
  lat: number;
  /** Cue start time in seconds — lets the path line up with video playback. */
  t: number;
  /** Relative altitude in metres, when present. */
  alt: number | null;
}

export interface Bounds {
  /** [lon, lat] */
  min: [number, number];
  max: [number, number];
}

/**
 * Parse a cue's GPS to `[lon, lat]`, or `null` when it's missing or a
 * null-island `(0, 0)` fix (DJI writes that before the aircraft has a lock).
 */
export function parsePosition(cue: Cue): [number, number] | null {
  const lat = Number(cue.data.latitude);
  const lon = Number(cue.data.longitude);
  if (!Number.isFinite(lat) || !Number.isFinite(lon)) return null;
  if (lat === 0 && lon === 0) return null;
  if (Math.abs(lat) > 90 || Math.abs(lon) > 180) return null;
  return [lon, lat];
}

/** Extract the ordered list of located points from a clip's cues. */
export function extractTrack(cues: readonly Cue[]): TrackPoint[] {
  const out: TrackPoint[] = [];
  for (const cue of cues) {
    const pos = parsePosition(cue);
    if (!pos) continue;
    const alt = Number(cue.data.rel_alt);
    out.push({
      lon: pos[0],
      lat: pos[1],
      t: cue.start,
      alt: Number.isFinite(alt) ? alt : null,
    });
  }
  return out;
}

/** Bounding box of a track, or `null` when it has no points. */
export function trackBounds(track: readonly TrackPoint[]): Bounds | null {
  if (track.length === 0) return null;
  let minLon = Infinity;
  let minLat = Infinity;
  let maxLon = -Infinity;
  let maxLat = -Infinity;
  for (const p of track) {
    if (p.lon < minLon) minLon = p.lon;
    if (p.lat < minLat) minLat = p.lat;
    if (p.lon > maxLon) maxLon = p.lon;
    if (p.lat > maxLat) maxLat = p.lat;
  }
  return { min: [minLon, minLat], max: [maxLon, maxLat] };
}

/** GeoJSON LineString coordinates (`[lon, lat][]`) for the path. */
export function lineCoordinates(track: readonly TrackPoint[]): [number, number][] {
  return track.map((p) => [p.lon, p.lat]);
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Telemetry/FlightPath.swift
// Pure helpers to turn parsed DJI telemetry into a flight path — no map
// library. Port of `src/shared/telemetry/flight-path.ts`. Extracts the GPS
// track from the cues `parseSrt` produced, so the same data that drives the
// telemetry panels draws the map.
//
// The web's `[lon, lat]` pairs are a `LonLat` here, the order kept in the
// name; its `Bounds` is `TrackBounds`, so nothing generic sits at the top of
// the module.

import Foundation

/// A GeoJSON position: longitude first, as the web's `[lon, lat]`.
public struct LonLat: Equatable, Sendable {
    public var lon: Double
    public var lat: Double

    public init(lon: Double, lat: Double) {
        self.lon = lon
        self.lat = lat
    }

    public init(_ lon: Double, _ lat: Double) {
        self.lon = lon
        self.lat = lat
    }
}

public struct TrackPoint: Equatable, Sendable {
    public var lon: Double
    public var lat: Double
    /// Cue start time in seconds — lets the path line up with video playback.
    public var t: Double
    /// Relative altitude in metres, when present.
    public var alt: Double?

    public init(lon: Double, lat: Double, t: Double, alt: Double?) {
        self.lon = lon
        self.lat = lat
        self.t = t
        self.alt = alt
    }
}

/// The web's `Bounds`: a bounding box in `[lon, lat]`.
public struct TrackBounds: Equatable, Sendable {
    public var min: LonLat
    public var max: LonLat

    public init(min: LonLat, max: LonLat) {
        self.min = min
        self.max = max
    }
}

private func finiteNumber(_ value: String?) -> Double? {
    guard let value, let n = Double(value), n.isFinite else { return nil }
    return n
}

/// Parse a cue's GPS, or nil when it is missing or a null-island `(0, 0)` fix
/// (DJI writes that before the aircraft has a lock).
public func parsePosition(_ cue: Cue) -> LonLat? {
    guard let lat = finiteNumber(cue.data["latitude"]), let lon = finiteNumber(cue.data["longitude"]) else {
        return nil
    }
    if lat == 0 && lon == 0 { return nil }
    if abs(lat) > 90 || abs(lon) > 180 { return nil }
    return LonLat(lon: lon, lat: lat)
}

/// Extract the ordered list of located points from a clip's cues.
public func extractTrack(_ cues: [Cue]) -> [TrackPoint] {
    var out: [TrackPoint] = []
    for cue in cues {
        guard let pos = parsePosition(cue) else { continue }
        out.append(TrackPoint(lon: pos.lon, lat: pos.lat, t: cue.start, alt: finiteNumber(cue.data["rel_alt"])))
    }
    return out
}

/// Bounding box of a track, or nil when it has no points.
public func trackBounds(_ track: [TrackPoint]) -> TrackBounds? {
    if track.isEmpty { return nil }
    var minLon = Double.infinity
    var minLat = Double.infinity
    var maxLon = -Double.infinity
    var maxLat = -Double.infinity
    for p in track {
        if p.lon < minLon { minLon = p.lon }
        if p.lat < minLat { minLat = p.lat }
        if p.lon > maxLon { maxLon = p.lon }
        if p.lat > maxLat { maxLat = p.lat }
    }
    return TrackBounds(min: LonLat(lon: minLon, lat: minLat), max: LonLat(lon: maxLon, lat: maxLat))
}

/// GeoJSON LineString coordinates (`[lon, lat][]`) for the path.
public func lineCoordinates(_ track: [TrackPoint]) -> [LonLat] {
    track.map { LonLat(lon: $0.lon, lat: $0.lat) }
}
